### PR TITLE
CA-271406 don't call infer_vgpu_map when VM is not live

### PR DESCRIPTION
**Pull request by lindig**
During VM migration, a VM may shut down. When this happens, the VM's
vGPUs get released immediately. This needs to be detected before
infer_vgpu_map() is called as it relies on vGPUs being assigned.

This commit moves the call to infer_vgpu_map into the try-catch block
that deals with edge cases like this. Before infer_vgpu_map is called we
check that the VM is live and hence has its vGPUs still assigned. To
avoid a data race, the liveness check is done only after no more
events from xenopsd are accepted.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

### CODE REVIEW

**Comment by lindig** [comment]
When accepted, this needs also go into the master branch.

**Comment by jonludlam** [comment]
This looks good to me. Local exception declaration, woo!

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -211,8 +211,11 @@ let infer_vgpu_map ~__context ?remote vm =
     List.map (fun self ->
       let vgpu = Db.VGPU.get_record ~__context ~self in
       let device = vgpu.API.vGPU_device in
+      let pgpu = vgpu.API.vGPU_scheduled_to_be_resident_on in
+      if pgpu = Ref.null then (* the VM was halted/suspended *)
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: Don't do this - do Db.is_valid_ref instead.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -211,8 +211,11 @@ let infer_vgpu_map ~__context ?remote vm =
...
+        raise (Xenops_interface.Does_not_exist("PGPU", Ref.string_of vm));
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: and I don't like the idea of raising Xenops_interface errors - that's quite misleading.

Comment (jonludlam) [reply ocaml/xapi/xapi_vm_migrate.ml]: Also, it'd be good to have a comment here saying precisely _why_ we're handling this carefully.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 3]
```diff
@@ -224,8 +235,12 @@ let infer_vgpu_map ~__context ?remote vm =
     List.map (fun self ->
       let vgpu = XenAPI.VGPU.get_record rpc session self in
       let device = vgpu.API.vGPU_device in
+      let pgpu = vgpu.API.vGPU_scheduled_to_be_resident_on in
+      if not (Db.is_valid_ref __context pgpu) then
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: This won't work. Hmm, I don't think we expose 'is_valid_ref' externally - perhaps we could emulate by just doing a get_uuid or something?

**Comment by lindig** [comment]
@jonludlam I've cleaned up the commits and simplified the fix. Could you take another look?

**Comment by jonludlam** [comment]
OK, looks good now.

**Pull request merged**